compiler/x/codegen/OMRTreeEvaluator.cpp

```diff
@@ -1578,6 +1590,185 @@ TR::Register *OMR::X86::TreeEvaluator::REPArraycmpLenEvaluator(TR::Node *node, T
    generateLabelInstruction(JMP4, node, endResultLabel, cg);
 
    //Result 
+   cg->stopUsingRegister(src1AddrRegister);
+   cg->stopUsingRegister(src2AddrRegister);
+   deps->stopAddingConditions();
+
+   generateLabelInstruction(LABEL, node, endResultLabel, deps, cg);
+   endResultLabel->setEndInternalControlFlow();
+
+   node->setRegister(resultRegister);
+
+   cg->decReferenceCount(src1AddrNode);
+   cg->decReferenceCount(src2AddrNode );
+   cg->decReferenceCount(lengthNode );
+
+   return resultRegister;
+   }
+
+TR::Register *OMR::X86::TreeEvaluator::ConstLenArraycmpEvaluator(TR::Node *node, TR::CodeGenerator *cg, uintptrj_t arrLen)
+   {
+   TR::Register *resultRegister = cg->allocateRegister(TR_GPR);
+   TR::Node *src1AddrNode = node->getChild(0);
+   TR::Node *src2AddrNode = node->getChild(1);
+   TR::Node *lengthNode = node->getChild(2);
+   
+   int32_t byteLen = 4;
+   if (TR::Compiler->target.is64Bit())
+      byteLen = 8;
+
+   TR::Register *src1AddrRegister = TR::TreeEvaluator::intOrLongClobberEvaluate(src1AddrNode, TR::TreeEvaluator::getNodeIs64Bit(src1AddrNode, cg), cg);
+   TR::Register *src2AddrRegister = TR::TreeEvaluator::intOrLongClobberEvaluate(src2AddrNode, TR::TreeEvaluator::getNodeIs64Bit(src2AddrNode, cg), cg);
+
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *notEqualLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *lessThanLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endResultLabel = generateLabelSymbol(cg);
+
+   generateLabelInstruction(LABEL, node, startLabel, cg);
+   startLabel->setStartInternalControlFlow();
+
+   TR::RegisterDependencyConditions *deps = generateRegisterDependencyConditions((uint8_t) 0, 3, cg);
+   deps->addPostCondition(src1AddrRegister, TR::RealRegister::esi, cg);
+   deps->addPostCondition(src2AddrRegister, TR::RealRegister::edi, cg);
+   deps->addPostCondition(resultRegister, TR::RealRegister::NoReg, cg);
+
+   //Deal with the residual bits first, as we want to have a length that is a multiple of 8 in order to use CMPSQ
+   uintptrj_t residualLen = arrLen % 8;
+   
+   if (residualLen == 7 || residualLen == 6 || residualLen == 5 || residualLen == 4)
```

Comment:
could you check bits for these conditions?

isn't the first condition just residualLen & 4, the second residualLen & 2, and the third residualLen & 1 ?